File: sudoku-js/src/test/js/sudoku/SolverStrategyCompositeTest.js
console && console.log("### SolverStrategyCompositeTest.js begin");

//TODO add tests of applyOnce()

(function(ns,test,lang,tools)
{
    const LOG = Logger.getLogger("SolverStrategyCompositeTest");
    ns.solverStrategyCompositeTestSuite = new test.TestSuite({name: "SolverStrategyCompositeTest"});


    let printBoard = function(board, hits)
    {
        if (LOG._isDebugOn) {
            LOG.debug(hits + " hits");
            LOG.debug(ns.SudokuHelper.printBoardDigits(board, '.'));
            LOG.debug(ns.SudokuHelper.printBoardDigitFlags(board));
        }
    };


//    public void test(final Board board) throws BoardSanityException
    let testBoard = function(board)
    {
        printBoard(board, 0);
        const solverStrategy = new ns.SolverStrategyComposite();
        const stopWatch = new tools.StopWatch().start();
        solverStrategy.apply(board);
        LOG.info("Solver elapsed time (ms): " + stopWatch.stop().elapsed());
    };


//    public void testBoardSolution(final Board board,final Board solutionBoard) throws BoardSanityException
    let testBoardSolution = function(board,solutionBoard)
    {
        board.setSolutionBoard(solutionBoard);
        testBoard(board);
    };


    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_01", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_01),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_01_SOLUTION));
    }));

    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_10", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_10),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_10_SOLUTION));
    }));

    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_20", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_20),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_20_SOLUTION));
    }));

    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_30", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_30),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_30_SOLUTION));
    }));

//    @Test(expected=BoardSanityException.class)
    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_40", active: true, expectedExceptionType: ns.BoardSanityException.TYPE}, function()
    {
        testBoard(ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_40));
    }));


    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_X0", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X0),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X0_SOLUTION));
    }));


    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_X1", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X1),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X1_SOLUTION));
    }));


    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_X2", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X2),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X2_SOLUTION));
    }));


    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_X3", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X3),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X3_SOLUTION));
    }));


    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_X4", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X4),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X4_SOLUTION));
    }));


    ns.solverStrategyCompositeTestSuite.addTest(new test.TestCase({name: "test_X5", active: true, expectedExceptionType: null}, function()
    {
        testBoardSolution(
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X5),
                ns.SudokuHelper.populateBoardFromText( new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_X5_SOLUTION));
    }));

}(Module("sudoku"),Module("test"),Module("lang"),Module("tools")));

console && console.log("### SolverStrategyCompositeTest.js end");

File: sudoku-j/src/test/java/com/dmarchesseault/sudoku/SolverStrategyCompositeTest.java
package com.dmarchesseault.sudoku;

import com.dmarchesseault.tools.StopWatch;
import org.apache.log4j.Logger;
import org.junit.Test;

//TODO add tests of applyOnce()

//todo impl with asserts
public class SolverStrategyCompositeTest
{
    private static final Logger LOG = Logger.getLogger(SolverStrategyCompositeTest.class);

    private void printBoard(final Board board, int hits)
    {
        if(LOG.isDebugEnabled())
        {
            LOG.debug(hits + " hits");
            LOG.debug(SudokuHelper.printBoardDigits(board, '.'));
            LOG.debug(SudokuHelper.printBoardDigitFlags(board));
        }
    }

    public void testBoard(final Board board) throws BoardSanityException
    {
        printBoard(board, 0);
        final SolverStrategy solverStrategy = new SolverStrategyComposite();

        final StopWatch stopWatch = new StopWatch().start();
        solverStrategy.apply(board);
        if(LOG.isDebugEnabled()) LOG.debug("Solver elapsed time (ms): " + stopWatch.stop().elapsed());
    }


    public void testBoardSolution(final Board board, final Board solutionBoard) throws BoardSanityException
    {
        board.setSolutionBoard(solutionBoard);
        testBoard(board);
    }

    @Test
    public void test_01() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_01),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_01_SOLUTION));
    }

    @Test
    public void test_10() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_10),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_10_SOLUTION));
    }

    @Test
    public void test_20() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_20),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_20_SOLUTION));
    }

    @Test
    public void test_30() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_30),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_30_SOLUTION));
    }

    @Test(expected=BoardSanityException.class)
    public void test_40() throws BoardSanityException
    {
        testBoard(SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_40));
    }

    @Test
    public void test_X0() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X0),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X0_SOLUTION));
    }

    @Test
    public void test_X1() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X1),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X1_SOLUTION));
    }

    @Test
    public void test_X2() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X2),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X2_SOLUTION));
    }

    @Test
    public void test_X3() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X3),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X3_SOLUTION));
    }

    @Test
    public void test_X4() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X4),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X4_SOLUTION));
    }

    @Test
    public void test_X5() throws BoardSanityException
    {
        testBoardSolution(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X5),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_X5_SOLUTION));
    }

}
